From e9ed23f84eb498d585ec2696903524fe9fa6e214 Mon Sep 17 00:00:00 2001
From: demostanis <demostanis@protonmail.com>
Date: Sat, 24 Sep 2022 17:17:57 +0200
Subject: [PATCH] Remove annoying AltGrrrr+Space behavior

---
 lafayette_linux_v0.6.1.py | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/lafayette_linux_v0.6.1.py b/lafayette_linux_v0.6.1.py
index 5fcb3a7..36c2b5c 100644
--- a/lafayette_linux_v0.6.1.py
+++ b/lafayette_linux_v0.6.1.py
@@ -119,7 +119,7 @@ LAYOUTS = {'fr': [{
             key <LSGT> {[ less            , greater         , lessthanequal   , greaterthanequal, VoidSymbol      , VoidSymbol      , VoidSymbol      , VoidSymbol      ]}; // < > ≤ ≥
 
             // Space bar
-            key <SPCE> {[ space           , nobreakspace    , U2019           , U2019           , nobreakspace    , U202F           , VoidSymbol      , VoidSymbol      ]}; //     ’ ’
+            key <SPCE> {[ space           , nobreakspace    , U2019           , U2019           , nobreakspace    , U2019           , VoidSymbol      , VoidSymbol      ]}; //     ’ ’
 
             // The “OneDeadKey” is an ISO_Level3_Latch, i.e. a “dead AltGr” key:
             // this is the only way to have a multi-purpose dead key with XKB.
@@ -238,7 +238,7 @@ LAYOUTS = {'fr': [{
             key <LSGT> {[ less            , greater         , VoidSymbol      , VoidSymbol      , VoidSymbol      , VoidSymbol      , VoidSymbol      , VoidSymbol      ]}; // < >
 
             // Space bar
-            key <SPCE> {[ space           , nobreakspace    , U2019           , U2019           , nobreakspace    , U202F           , VoidSymbol      , VoidSymbol      ]}; //     ’ ’
+            key <SPCE> {[ space           , nobreakspace    , U2019           , U2019           , nobreakspace    , U2019           , VoidSymbol      , VoidSymbol      ]}; //     ’ ’
 
             // The “OneDeadKey” is an ISO_Level3_Latch, i.e. a “dead AltGr” key:
             // this is the only way to have a multi-purpose dead key with XKB.
-- 
2.37.3

