--- src/modules/keyboard_state.cpp.orig	2022-04-28 00:30:12.783551214 +0200
+++ src/modules/keyboard_state.cpp	2022-04-28 00:42:02.768538022 +0200
@@ -189,9 +189,7 @@
     const std::string& format;
     const char* name;
   } label_states[] = {
-      {(bool)numl, numlock_label_, numlock_format_, "Num"},
       {(bool)capsl, capslock_label_, capslock_format_, "Caps"},
-      {(bool)scrolll, scrolllock_label_, scrolllock_format_, "Scroll"},
   };
   for (auto& label_state : label_states) {
     std::string text;
@@ -200,9 +198,10 @@
                        fmt::arg("name", label_state.name));
     label_state.label.set_markup(text);
     if (label_state.state) {
+      event_box_.show();
       label_state.label.get_style_context()->add_class("locked");
     } else {
-      label_state.label.get_style_context()->remove_class("locked");
+      event_box_.hide();
     }
   }
 
